File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/NotificationService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Notification;

import java.util.Set;

public interface NotificationService {

    /**
     * Save a notification
     *
     * @param notification to save
     * @return the notification
     */
    Notification createNotification(Notification notification);

    /**
     * Update a notification
     *
     * @param notification to update
     * @return the updated notification
     */
    Notification updateNotification(Notification notification);

    /**
     * Get all notifications for a recipient
     *
     * @param id of the profile
     * @return Set of all notifications sent to a profile
     */
    Set<Notification> findAllByRecipient_Id(Long id);

    /**
     * Get all applications for an event
     *
     * @param id of the event
     * @return Set of all Notifications with type application for an event
     */
    Set<Notification> findAllApplicationsByEvent_Id(Long id);

    /**
     * Delete a notification
     *
     * @param id to delete
     * @param authorization used for authorizing who can delete the notification
     */
    void deleteNotification(Long id, String authorization);

    /**
     * Change (toggle) favourite of Notification (true <-> false)
     *
     * @param notification to change Boolean favorite
     * @return Notification with changed favourite
     */
    Notification changeFavorite(Notification notification);

    /**
     * Finds the only notification sent by a user regarding an event
     *
     * @param eventId id of event
     * @param employeeId id of the employee who sent the notification
     * @return the notification sent by this employee profile with the given id regarding this event
     */
    Notification findFirstByEvent_IdAndSender_Id(Long eventId, Long employeeId);

    /**
     * Delete an application and its notification
     *
     * @param id of the notification to delete
     * @param authorization used for authorizing who can delete the notification
     */
    void deleteApplication(Long id, String authorization);

    /**
     * Delete an employee from assigned task
     *
     * @param taskId task that employee quit
     * @param authorization token to identify logged in user (employee)
     */
    void deleteEmployeeFromTask(Long taskId, String authorization);
}

File: frontend/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Globals} from '../global/globals';
import {Observable} from 'rxjs';
import {SimpleNotification} from '../dtos/simple-notification';

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private notificationBaseUri: string = this.globals.backendUri + '/notifications';
  constructor(private httpClient: HttpClient, private globals: Globals) { }

  /**
   * Gets all notifications from the backend
   */
  getNotifications(): Observable<SimpleNotification[]> {
    console.log('GET notifications');
    return this.httpClient.get<SimpleNotification[]>(this.notificationBaseUri);
  }

  /**
   * Deletes the notification with the given id
   * @param id of the notification to delete
   */
  deleteNotification(id: number) {
    console.log('Delete notification with id: ' + id);
    return this.httpClient.delete(this.notificationBaseUri + '/' + id);
  }

  /**
   * Updates the notification
   * @param notification with the updated values
   */
  updateNotification(notification: SimpleNotification) {
    console.log('Update notification: {}', notification);
    return this.httpClient.put<SimpleNotification>(this.notificationBaseUri + '/' + notification.id, notification);
  }

  /**
   * Changes the favourite boolean param of a notification
   * @param notification with the updated favourite value
   */
  changeFavorite(notification: SimpleNotification) {
    console.log('change notification.favorite: {}', notification.favorite);
    return this.httpClient.put<SimpleNotification>(this.notificationBaseUri + '/changeFavorite', notification);
  }
}
